include/OscillatorConstants.h

```diff
@@ -0,0 +1,52 @@
+/*
+ * OscillatorConstants.h - declaration of constants used in  Oscillator and SampleBuffer
+ * for band limited wave tables
+ *
+ * Copyright (c) 2018      Dave French	<dave/dot/french3/at/googlemail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#ifndef OSCILLATORCONSTANTS_H
+#define OSCILLATORCONSTANTS_H
+
+#include <array>
+
+#include "lmms_basics.h"
+
+namespace OscillatorConstants
+{
+	// Limit wavetables to the audible audio spectrum
+	const int MAX_FREQ = 20000;
+	// Minimum size of table to have all audible bands for midi note 1 (i.e. 20 000 Hz / 8.176 Hz)
+	constexpr int WAVETABLE_LENGTH = static_cast<int>(MAX_FREQ / 8.176);
+
+	//SEMITONES_PER_TABLE, the smaller the value the smoother the harmonics change on frequency sweeps
+	// with the trade off of increased memory requirements to store the wave tables
+	// require memory = NumberOfWaveShapes*WAVETABLE_LENGTH*(MidiNoteCount/SEMITONES_PER_TABLE)*BytePerSample_t
+	// 7*2446*(128/1)*4 = 8766464 bytes
+	const int SEMITONES_PER_TABLE = 1;
+	const int WAVE_TABLES_PER_WAVEFORM_COUNT = 128 / SEMITONES_PER_TABLE;
+
+	typedef std::array<sample_t, WAVETABLE_LENGTH> wavetable_t;
```

Comment:
Isn't it `WAVE_LENGTH`? I'd expect that a "table" is a set of waves, and this is just one wave? In this case, `wavetable_t` does not make much sense, either, it should be `wavetable_entry_t`, and `waveform_t` should be `wavetable_t`. Or am I confusing this all?

Reply:
I thought about that one for a while when deciding on the names. The problem is that "wave table" is a term that has been quite "muddied" by years of misuse by marketing departments.

I originally understood is the same way as you describe, that "wavetable" is a table of waveforms, each shaped differently. That understanding comes from modern "wavetable synthesis", which is sold as "sample-based synthesis, but with dynamic switching / modulating / distorting of the sample on the fly", where the "wave table" would be a "table of available samples / waveforms". Or something like that, I'm probably mixing a lot of terms here.

The original meaning of wave table (the one the original PR used, and the one I kept) is simply "a table that contains values that describe a wave". So basically a lookup table, typically containing one cycle of the wave. Wavetable synthesis then refers to any synthesizer that uses wave table lookups as the source of signal, instead of generating the samples on the fly. So when you set TripleOsc (before this PR) to Sine wave, it is a normal synthesizer, but it you load a sine wave to the user waveform, it is a wavetable synthesizer. Magic. :stars: 

It is quite well explained at: https://en.wikipedia.org/wiki/Wavetable_synthesis Before reading this, I _knew for sure_ that wavetable is a set of waves. But I hate abusive marketing more than I hate being wrong, so I changed my mind. :smile: 

Anyway; `wavetable_t` being set in stone (because it describes the smallest element, a table describing one period of a wave), I had to find some other term for a set of wave tables. "Waveform" seemed best to me: it describes basically the same, but it feels a bit more abstract and generic. I.e.:
 - I have two wave _shapes_: a sine wave and square wave (describing how the signal looks in general).
 - Each shape is represented as a digital _waveform_ ("form" is close in meaning to "shape", but also closer to computer terminology used to digitally stored sound waves).
 - Each waveform is then implemented as a series of _wave tables_ with varying frequency content, each made to best represent the waveform or shape at a specific pitch (lowest level, describes the specific implementation).

That is my best guess, but I wouldn't be surprised if I'm understanding something incorrectly and there is a better set of terms.

Reply:
OK, I got it. They should have better named it "wavearry" synth or "wavevector" synth.

Maybe you should add a comment above `wavetable_t` or `WAVETABLE_LENGTH`, describing what you actually consider a wavetable?

Btw: We currently add a modulation through a table of waves into zyn, and you can tell zyn that the table is generated with different parameters. E.g. you can switch from a power wave to a saw wave by modulating the basewave parameter, and then zyn internally cycles through all the generated waves by altering the index of the table of waves. I called that whole thing "wavetable modulation". I think that term should still be accurate given the other definition that you propose in this PR, right?

Reply:
> Maybe you should add a comment above `wavetable_t` or `WAVETABLE_LENGTH`, describing what you actually consider a wavetable?

Good idea, will do.
 
> Btw: We currently add a modulation through a table of waves into zyn, and you can tell zyn that the table is generated with different parameters. E.g. you can switch from a power wave to a saw wave by modulating the basewave parameter, and then zyn internally cycles through all the generated waves by altering the index of the table of waves. I called that whole thing "wavetable modulation". I think that term should still be accurate given the other definition that you propose in this PR, right?

Well, if I learned anything from the Wikipedia article, it is that any term is accurate if you know how to sell it. :smile: As I understand it, what you describe is technically correct, although I think that the specific feature that was important for "modern wavetable synthesizers" (i.e. ~1970, switching through a set of waves) was that there was some interpolation between adjacent waves in the set (so there are no abrupt changes in timbre when you sweep / modulate through them slowly).

So if the new feature in Zyn lets you set the basewave parameter to 0.5 and have a 50 % power, 50 % saw wave mix, that looks to me very close "modern wavetable synthesis", but with the wavetable fixed to predefined shapes and not user-selectable. If the basewave selector moves only by integer steps, it would be probably a bit harder to sell, but technically "wavetable modulation" probably still makes sense. :))

But I don't really know, I'm not an expert on the terminology, I just googled until I found a local minimum in my confusion and picked a few terms that would make sense in the Oscillator class. :smile:

Reply:
OK, feel free to close this comment after you added the comment in the code.